### PR TITLE
`Development`: Increase client test coverage for complaints

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).

### Description

Added some tests and made some small code improvements.
I didn't add too many tests to `list-of-complaints.component.ts` because a lot of them would've been just functionality reproduction. Let me know if you think I missed suitable testcases there.

#### Code Review
- [x] Review 1
- [x] Review 2

### Test Coverage

All: 84.23% | 88.47%
 
complaint-response.service.ts: 77.41% | 100%
complaint.service.ts: 76.31% | 96.22%
complaints-student.view.component.ts: 78.57% | 95.89%
complaints-for-tutor.component.ts: 67.3% | 82.47%
complaints-form.component.ts: 85.71% | 100%
list-of-complaints.component.ts: 81.25% | 93.4%
complaint-request.component.ts: 100% | 75%
complaint-response.component.ts: 100% | 75%

### CODE REVIEW

**Review by b-fein — DISMISSED** [review]
The new tests look good. :+1:
I added a few inline comments about some small improvements. I also added a suggestion for potential additional tests for `list-of-complaints`.

File: src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            });
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    describe('loadComplaints', () => {
+        it('find for tutor by exercise', () => {
+            comp.tutorId = 12;
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts]: Done, thanks for the suggestion

File: src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts:1]: I merged them, thanks.

File: src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                id: 42,
+                submittedTime: dayjs().subtract(8, 'days'),
+            } as Complaint;
+
+            const result = comp.shouldHighlightComplaint(complaint);
+
+            expect(result).toBe(true);
+        });
+    });
+
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts:215]: Thanks for the suggestions, for the first one I didn't want to test it as it uses translations but I forgot I could just spy on it and verify that ^^

File: src/test/javascript/spec/component/complaints/complaint-response.service.spec.ts [hunk 4]
```diff
@@ -12,29 +12,31 @@ import { User } from 'app/core/user/user.model';
...
-    let testBed: TestBed;
     let complaintResponseService: ComplaintResponseService;
     let httpTestingController: HttpTestingController;
     let defaultComplaintResponse: ComplaintResponse;
+    let accountService: AccountService;
     let expectedComplaintResponse: any;
 
     beforeEach(() => {
-        TestBed.configureTestingModule({
+        return TestBed.configureTestingModule({
```
Comment (b-fein) [inline src/test/javascript/spec/component/complaints/complaint-response.service.spec.ts:22]:
```suggestion
        TestBed.configureTestingModule({
```
The `return` should’t be needed here, I think.

File: src/test/javascript/spec/component/complaints/complaint.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    clientComplaint2.id = 42;
+    clientComplaint2.complaintType = ComplaintType.MORE_FEEDBACK;
+    clientComplaint2.result = new Result();
+    clientComplaint2.submittedTime = dayjsTime2;
+    clientComplaint2.complaintText = 'Another test text';
+
+    const serverComplaint2 = { ...clientComplaint2, submittedTime: stringTime2 };
+
+    beforeEach(() => {
+        return TestBed.configureTestingModule({
```
Comment (b-fein) [inline src/test/javascript/spec/component/complaints/complaint.service.spec.ts:42]:
```suggestion
        TestBed.configureTestingModule({
```
As in the other file.

File: src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        student: {},
+    } as Complaint;
+    const complaint5 = {
+        id: 5,
+        accepted: false,
+        student: {},
+    } as Complaint;
+
+    beforeEach(() => {
+        return TestBed.configureTestingModule({
```
Comment (b-fein) [inline src/test/javascript/spec/component/complaints/list-of-complaints.component.spec.ts:66]:
```suggestion
        TestBed.configureTestingModule({
```

File: src/main/webapp/app/complaints/list-of-complaints/list-of-complaints.component.ts [hunk 7]
```diff
@@ -30,7 +30,7 @@ export class ListOfComplaintsComponent implements OnInit {
     private exerciseId: number;
     private tutorId: number;
     private examId?: number;
-    private correctionRound?: number;
+    correctionRound?: number;
```
Comment (b-fein) [inline src/main/webapp/app/complaints/list-of-complaints/list-of-complaints.component.ts:33]:
```suggestion
    private correctionRound?: number;
```
I don’t think this is accessed directly in the tests any more and can be set to `private`.

**Review by b-fein — APPROVED** [review]
The new additional tests look good, too. :+1: I only found some very small (possible) improvements on the re-review.

File: src/main/webapp/app/complaints/list-of-complaints/list-of-complaints.component.ts (hunk 7 above)
Comment (julian-christl) [reply src/main/webapp/app/complaints/list-of-complaints/list-of-complaints.component.ts:33]: It is, `list-of-complaints.component.spec.ts:325`

**Review by b-fein — APPROVED** [review]

**Review by willergeorg — APPROVED** [review]
Newly added test cases look good to me. 👍

**Pull request merged**